File: app/src/main/java/co/k2/newsbits/dagger/AppComponent.java
package co.k2.newsbits.dagger;

import javax.inject.Singleton;

import co.k2.newsbits.App;
import co.k2.newsbits.data.DataModule;
import dagger.BindsInstance;
import dagger.Component;
import dagger.android.AndroidInjector;
import dagger.android.support.AndroidSupportInjectionModule;

/**
 * Copyright (C) 2019 K2 CODEWORKS
 * All rights reserved
 *
 * @author Kislay
 * @since 28/03/19
 */

@Singleton
@Component(modules = {
        AndroidSupportInjectionModule.class,
        AppModule.class,
        ActivityBindingModule.class,
        DataModule.class
})
public interface AppComponent extends AndroidInjector<App> {

    @Component.Builder
    interface Builder {

        @BindsInstance
        AppComponent.Builder application(App application);

        AppComponent build();

    }

}

File: app/src/main/java/co/k2/newsbits/dagger/AppComponent.kt
package co.k2.newsbits.dagger

import co.k2.newsbits.App
import co.k2.newsbits.headlines.di.HeadlinesComponent
import dagger.BindsInstance
import dagger.Component
import dagger.android.AndroidInjectionModule
import javax.inject.Singleton

@Singleton
@Component(
    modules = [
        AndroidInjectionModule::class, NetworkModule::class, ViewModelBuilderModule::class
    ]
)
interface AppComponent {

    @Component.Builder
    interface Builder {

        fun build(): AppComponent
        fun networkModule(networkModule: NetworkModule): Builder

        @BindsInstance
        fun application(application: App): Builder
    }

    fun headlinesComponent(): HeadlinesComponent.Factory

}
